### PR TITLE
Add annotation processor to save Javadoc

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: drift-javadoc/src/test/java/io/airlift/drift/javadoc/TestJavdocIntegration.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        assertGenerated(basedir, "Response");
+        assertGenerated(basedir, "SimpleLogger");
+    }
+
+    private static void assertGenerated(File basedir, String name)
+            throws IOException
+    {
+        URL expected = getResource(format("basic/%s.txt", name.replace("$", "")));
+
+        name += "$driftMeta";
```
Comment (electrum) [inline drift-javadoc/src/test/java/io/airlift/drift/javadoc/TestJavdocIntegration.java]: Use constant

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import static javax.lang.model.element.ElementKind.ENUM_CONSTANT;
+import static javax.tools.Diagnostic.Kind.ERROR;
+import static javax.tools.Diagnostic.Kind.NOTE;
+import static javax.tools.Diagnostic.Kind.WARNING;
+
+@SupportedAnnotationTypes({
+        ThriftAnnotations.THRIFT_ENUM,
+        ThriftAnnotations.THRIFT_SERVICE,
+        ThriftAnnotations.THRIFT_STRUCT})
+@SupportedSourceVersion(SourceVersion.RELEASE_8)
```
Comment (dain) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: maybe use latest so it works in 9?

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        TypeSpec.Builder typeSpec = TypeSpec.classBuilder(className + META_SUFFIX)
+                .addAnnotation(documentationAnnotation(serviceDocumentation));
+
+        // offset auto-generated order numbers so they don't collide with hand-written ones
+        AtomicInteger nextOrder = new AtomicInteger(10000);
+
+        for (Element element : elements().getAllMembers(typeElement)) {
+            String name = element.getSimpleName().toString();
+            List<String> comment = getComment(element);
+            if ((element instanceof ExecutableElement) &&
```
Comment (dain) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: one line maybe?

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/ThriftAnnotations.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+final class ThriftAnnotations
+{
+    public static final String THRIFT_DOCUMENTATION = ThriftDocumentation.class.getName();
+    public static final String THRIFT_ORDER = ThriftOrder.class.getName();
+    public static final String THRIFT_FIELD = ThriftField.class.getName();
+    public static final String THRIFT_METHOD = ThriftMethod.class.getName();
+
+    public static final String THRIFT_ENUM = "io.airlift.drift.annotations.ThriftEnum";
+    public static final String THRIFT_SERVICE = "io.airlift.drift.annotations.ThriftService";
+    public static final String THRIFT_STRUCT = "io.airlift.drift.annotations.ThriftStruct";
```
Comment (dain) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/ThriftAnnotations.java]: Not sure why some of these are constant strings and some are class names

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/ThriftAnnotations.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    public static final String META_SUFFIX = "$driftMeta";
```
Comment (dain) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/ThriftAnnotations.java]: Uppercase `D` maybe?

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+import static javax.tools.Diagnostic.Kind.ERROR;
+import static javax.tools.Diagnostic.Kind.NOTE;
+import static javax.tools.Diagnostic.Kind.WARNING;
+
+@SupportedAnnotationTypes({
+        ThriftAnnotations.THRIFT_ENUM,
+        ThriftAnnotations.THRIFT_SERVICE,
+        ThriftAnnotations.THRIFT_STRUCT})
+@SupportedSourceVersion(SourceVersion.RELEASE_8)
+public class JavadocProcessor
```
Comment (dain) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: Maybe we should name this `DocumentationExtractor`?  Just and idea

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/ThriftAnnotations.java (hunk 4 above)
Comment (electrum) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/ThriftAnnotations.java]: The constants are necessary to be used as annotation arguments for the annotation processor

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java (hunk 2 above)
Comment (electrum) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: That's an enum, we'd need to compile against JDK 9 for that

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java (hunk 6 above)
Comment (electrum) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: Fine with me. Should we also rename the module? `drift-documentation-extractor`?

Comment (electrum) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: I'm going to leave this for now, we can rename later.

File: drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java (hunk 3 above)
Comment (electrum) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: I tried that and the line length combined with the `&&` and `||` makes it harder to read

Comment (electrum) [inline drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java]: There were some unnecessary parens on here that made it look weird. Removed.

**Pull request merged**